### DISCUSSION TITLE
Consume non searchable courses

### DISCUSSION THREAD

**Discussion by kiranharidas187** [Sunbird-Ed/Community#398] [Bugs] [answered]
@vinukumar-vs @Krishnaj20  We have a use-case where a course could be added to the improvement project and the user could access the course only through the improvement project and not through course search in the platform. 

**Current Process followed**

- creates a course / Copy a course from the platform. Create. a batch for the course 
- Send the course id to the implementation team to unlist
- Implementation team calls following API https://preprod.ntp.net.in/action/content/v3/unlisted/publish/COURSE_ID
- Map course to program and allows user to consume

**Problem**

After mapping the above-mentioned steps, users are not able to consume the course. Showing the below error while trying to join the course for the first time
<img width="688" alt="Screenshot 2023-02-14 at 12 37 15 PM" src="https://user-images.githubusercontent.com/21116904/218664171-42dd36ca-93b6-4f60-804f-ca176b65e9e0.png">

Second time trying to join the course again its showing below error 
API: https://preprod.ntp.net.in/learner/course/v1/enrol
Response:` {"id":"api.course.enroll","ver":"v1","ts":"2023-02-14 07:08:52:449+0000","params":{"resmsgid":null,"msgid":"a4bf54eb-b762-f27c-bc31-34252548c99e","err":"USER_ALREADY_ENROLLED_COURSE","status":"USER_ALREADY_ENROLLED_COURSE","errmsg":"User has already Enrolled this course ."},"responseCode":"CLIENT_ERROR","result":{}}`

**Note**
This flow was working fine in previous releases

Sample course which has this  issue: https://preprod.ntp.net.in/explore-course/course/do_2137130186269818881235

Please let us know what could be the cause of this issue

**Comment by Krishnaj20** [comment]
@kiranharidas187 - was this scenario validated and certified as part of ED 5.1.0 release?

@ashokreddy1208 @chitranshu-keshav

**Reply by kiranharidas187** [reply]
@Krishnaj20 This issue was identified in prod. The same is reproducible in 5.1.0 also.  Was trying to understand if there were any flow changes done recently.

**Reply by Krishnaj20** [reply]
@kiranharidas187 - was this scenario validated and certified as part of ED 5.1.0 on staging? Prod environment for ED is staging.

**Reply by kiranharidas187** [reply]
@Krishnaj20 This issue is present in the staging of ED and Diksha. But it was working as expected in the old release

**Reply by vinukumar-vs** [reply]
@kiranharidas187 I would suggest you to share the full log information of error taking it from Graylog. As you commented, User is already enrolled to the course.

Can you try to enroll to the new course & share the logs if the issue still persist. 

> API: https://preprod.ntp.net.in/learner/course/v1/enrol
Response: {"id":"api.course.enroll","ver":"v1","ts":"2023-02-14 07:08:52:449+0000","params":{"resmsgid":null,"msgid":"a4bf54eb-b762-f27c-bc31-34252548c99e","err":"USER_ALREADY_ENROLLED_COURSE","status":"USER_ALREADY_ENROLLED_COURSE","errmsg":"User has already Enrolled this course ."},"responseCode":"CLIENT_ERROR","result":{}}

@reshmi-nair @rhwarrier @Krishnaj20

**Reply by kiranharidas187** [reply]
I will try to get the log, but if the user has joined the course, the Join Course button should not be shown in the UI right? Its showing up now @vinukumar-vs

**Reply by kiranharidas187** [reply]
@vinukumar-vs @rhwarrier can we connect once and discuss this??

**Reply by rhwarrier** [reply]
please set up some time with @reshmi-nair and @vinukumar-vs

**Comment by kameshbhr** [comment]
@kiranharidas187 This looks like a workaround you are trying with the existing unlist option. The purpose of unlisted content is different. It was intended for testing/reviewing of the content through a consumption app, before making it live for everyone to use. So relying on this functionality to make courses consumable only through a link can result in many issues.
I suggest that you shouldn't follow this approach. The search and discoverability has to be a behavioral attribute of the content.

**Reply by kiranharidas187** [reply]
@kameshbhr I think we have got approval on this approach and hence implemented it. Can you guys confirm @aishwaryashikshalokam @Prateek-slokam ??

**Reply by aishwaryashikshalokam** [reply]
Dear @kameshbhr 

The flow and process to add courses to a Program were decided and concluded along with Diksha Team and DC team too.
This story was the part 4.3 sprint.

Here is the **solution note** of the same.  https://docs.google.com/document/d/1fBGw0clhVJjkE3l_08ePkcqd2vGJRfLGP5EU1x13zfo/edit?usp=sharing.

Attaching the **confluence link** too https://project-sunbird.atlassian.net/l/cp/81uQehPQ.

Can we please get on a call to discuss the issue and the gaps?

**Comment by kiranharidas187** [comment]
@kameshbhr @vinukumar-vs  can you guys help with this??

**Reply by Krishnaj20** [reply]
@kiranharidas187 - request you to set up some time this week with @kameshbhr @reshmi-nair @rhwarrier @vinukumar-vs & me to close this. We can do tomm 3 to 3:30, if it works for everyone.

**Reply by aishwaryashikshalokam** [reply]
Sent calendar invite for the same. 
Requesting everyone to make it possible to attend

**Comment by vinukumar-vs** [comment]
@kiranharidas187 @aishwaryashikshalokam 
Unlisted API's are deprecated in relese-5.2.0
https://knowlg.sunbird.org/use/deprecations/release-5.2.0#apis

@Krishnaj20 @kameshbhr @vrayulu @rhwarrier  @pallakartheekreddy @maheshkumargangula

**Reply by aishwaryashikshalokam** [reply]
Thank you team for coming together.

As discussed SL Engg team will test the same approach with new APIs and get back on the same.

**Reply by aks30** [reply]
@vinukumar-vs Looking for some help with the alternative APIs suggested. I have created a discussion forum for the same [here](https://github.com/orgs/Sunbird-Knowlg/discussions/116)

Cc - @aishwaryashikshalokam

**Reply by aks30** [reply]
@vinukumar-vs  We have tried the unlist API in Staging as suggested [here](https://github.com/orgs/Sunbird-Knowlg/discussions/116#discussioncomment-5508743) -  however the issue still remains.

Steps we have followed - 

1. In staging, using the content creator role user ([ContentCreatortn@yopmail.com](mailto:ContentCreatortn@yopmail.com))we have created a copy of the original course - [RootCourse-July14](https://staging.sunbirded.org/explore-course/course/do_21306399144626585613492)
2. Rename it to - "Copy of RootCourse-10April-Akash" and sent it for review
3. Published the copied course here - https://staging.sunbirded.org/learn/course/do_213771951428878336121
4. Unlisted the new course using unlist API, curl below
5. I am able to access the course using the direct link - https://staging.sunbirded.org/learn/course/do_213771951428878336121 but not available in search - https://staging.sunbirded.org/search/Library/1?key=Copy%20of%20RootCourse-10April-Akash&selectedTab=all
6. When a user tries to join the newly created course he gets the error mentioned in the [first comment here](https://github.com/orgs/Sunbird-Ed/discussions/398#discussion-4856186)

`curl --location --request POST 'https://staging.sunbirded.org/api/collection/v1/unlisted/publish/do_213771951428878336121' \
--header 'user-id: sunbird' \
--header 'Authorization: Bearer XXXXX' \
--header 'Content-Type: application/json' \
--data-raw '{
    "request": {
        "content": {
            "lastPublishedBy": "4cd4c690-eab6-4938-855a-447c7b1b8ea9"
        }
    }
}'`

**Reply by chitranshu-keshav** [reply]
@vinukumar-vs Awaiting your response on the above

**Reply by chitranshu-keshav** [reply]
@amitpriyadarshi can you please look into the above and get back

**Comment by amitpriyadarshi** [comment]
@aks30 as per your previous [comment](https://github.com/orgs/Sunbird-Ed/discussions/398#discussioncomment-5603624), I understood that you have  -

- Copied a course do_21306399144626585613492 `(RootCourse-July14)` to a new course do_213771951428878336121 `(Copy of RootCourse-10April-Akash)`
- Published the new copied course using the `unlisted/publish` api.
- Post that you are able to fetch course details using the direct link i.e. `https://staging.sunbirded.org/learn/course/do_213771951428878336121`
-  But while search the the course you are not able to fetch the course object

The reason behind this is: 
- This course exists in the system with the status `Unlisted` , i.e. Unlisted Publish (You can consider it as private publish)
- When we search for the course using name, in the Search api request payload, `status` is not passed.
- If the the search request does not contains `status`- by default `status = Live` is passed and fetches only those objects with status as `Live`
- Since this course status is `Unlisted`, it does not comes in response.

cc: @chitranshu-keshav @Krishnaj20 @kameshbhr @vinukumar-vs

**Reply by kameshbhr** [reply]
@amitpriyadarshi The issue is not that the course is not searchable. 
The issue is when a user is trying to join the course using the link, there is an error.

**Comment by amitpriyadarshi** [comment]
@aks30 can you explain and if possible connect to discuss the issue.

**Comment by amitpriyadarshi** [comment]
@aks30 @kiranharidas187 
I cross checked the reproducible step, existing work flow and code. 

- When the user try to join the course batch first time, he is enrolled to the batch. user_enrolments table entry confirm the same.
- But the enrolment list API does not return the course-batch against the user id, because, 
             * user enrolment list api fetches the course specific details using composite search API
             * in the search request `status=Live` is passed in filter.
             * because of this the course with `status=Unlisted` does not come in the response.
- Since the enrolment list does not contains this course batch, user get the option to enroll to that batch (again).
- But when the user try to enroll, a client error is returned with error `USER_ALREADY_ENROLLED_COURSE`

cc: @chitranshu-keshav @Krishnaj20 @kameshbhr @vinukumar-vs

**Reply by Krishnaj20** [reply]
Great analysis @amitpriyadarshi . Thank you!
This workflow used to work in older versions of the unlisted APIs prior to 5.2.0. Do you know why this is breaking now when prior to 5.2.0 it was working?

**Reply by amitpriyadarshi** [reply]
@Krishnaj20 unlisted published course will not come in enrolment list api. I verified in the previous releases, there also enrolment list api used to fetch only courses with status = Live, not  the courses with status= Unlisted.

**Reply by aks30** [reply]
@amitpriyadarshi This would be a bug right? As a user, logically I should be able to enroll in a private/unlisted course, correct? If not a bug, in case you can point to the corresponding story/PRD or a solution note which explains the behavior and implementation of unlisted courses for our better understanding

**Reply by Krishnaj20** [reply]
@aks30 - Unlisted content's purpose as explained by @kameshbhr  'The purpose of unlisted content is different. It was intended for testing/reviewing of the content through a consumption app, before making it live for everyone to use.'  So user not able to enrol into an unlisted course isn't a bug. 
But the hack that was deployed isn't working for the past 6-8 months - you may want to take help of @amitpriyadarshi & @vinukumar-vs to find out the root cause of this issue. Based on your findings we can discuss the next steps. 

@vrayulu @maheshkumargangula

**Reply by vrayulu** [reply]
This looks like a workaround way of unlisted APIs is used to enable the required feature. And some bugs in the implementation also helped this feature - enrolment into unlisted course is allowed but enrolment list is not returning unlisted courses.

A proper solution would have been to implement private courses and private batches. But the proper solution was never picked up and continued with short term solution for too long. I recommend implementing a proper solution this time rather than applying patches to the short term solution.

**Reply by aishwaryashikshalokam** [reply]
@vrayulu 
The short term solution was recommended by DC itself.  Since it is not functional anymore we will take the long term and permanent solution for the same.

As it involves courses and the solution may require enhancements in courses consumption flow, I'll suggest that the solutioning is done together along with the sunbird team.

Let's schedule calls for the same. We can pick up from the last solution note.

 @Krishnaj20 let me know suitable slots this week to connect on the same.

I am available on the following this week:

3rd May: 3:30-4:15/ 5:00-6:00pm
4th May : 12:00-1:00
5th May: 10:00-11:00

**Reply by aishwaryashikshalokam** [reply]
Hi @Krishnaj20 

Bringing this on top for you.

**Reply by Krishnaj20** [reply]
@aishwaryashikshalokam - we can meet on 16th May, 12 to 1 pm if it works for you. Please include @kameshbhr & @rhwarrier in the invite. Thanks

**Reply by aishwaryashikshalokam** [reply]
12:00-1:00 does not suit us on 16th May. Can we do anytime between 2:00-5:00pm.
We are also available on Wednesday between 10-12 and post 4.

**Reply by Krishnaj20** [reply]
@aishwaryashikshalokam - we can do 3pm to 4pm on 16th May.

**Comment by aishwaryashikshalokam** [comment]
Dear @kameshbhr and @Krishnaj20 

Thank you for your time today.  Given below are the discussion points:

- Behavior of courses that are part of the program shall be common irrespective of being original or copied.
- Any course that is part of the program will not be searchable in the library and will be accessible only via Program tile and link.
- The link should also ask the users to join the program just like other resources. Eg Projects, Observations, Survey
- The courses part of the Program will not show in Courses Tile.
- In the courses tile, only those courses will show up that are not part of any program
- Courses that need to be part of the program have to have a Program designer as their creator too

In Phase 2, we can also start brainstorming on how the data of courses part of programs also be reflected in Program Dashboard.

@vrayulu can we please discuss how the requirements mentioned above be implemented?

We will be happy to get on a call to discuss this and take it forward from a tech implementation perspective.

cc: @aks30 @kiranharidas187 @Prateek-slokam @vijiurs

**Reply by aishwaryashikshalokam** [reply]
@vrayulu any thoughts on the same?

We will be able to move forward only after hearing your input and feedback

**Reply by vrayulu** [reply]
@aishwaryashikshalokam - the above listed requirements for courses in a program looks fine to me. I would like to see the technical design for this and discuss the proposed design. Is there any tech design document or notes that can be shared?

**Reply by aishwaryashikshalokam** [reply]
Hi @vrayulu 

Thank you for your input. There is no technical design as of now for the same. We first wanted to see if the approach looks fine to you.

@kiranharidas187 @kameshbhr, now that it is approved by Rayulu, what shall be our next steps? How do you suggest we move forward?

**Reply by Krishnaj20** [reply]
@aishwaryashikshalokam - are you looking for any inputs from the community on this thread?

**Reply by aishwaryashikshalokam** [reply, marked as answer]
No @Krishnaj20 , I'll discuss it with my team internally and then reach out if anything is required.
